tests/Unit/Controllers/PublisherControllerTest.php

```diff
@@ -0,0 +1,316 @@
+<?php
+
+namespace Tests\Unit\Controllers;
+
+use App\Http\Controllers\Admin\PublisherController;
+use App\Http\Requests\PublisherRequest;
+use App\Models\Publisher;
+use App\Models\User;
+use App\Repositories\Publisher\PublisherRepositoryInterface;
+use Illuminate\Http\UploadedFile;
+use Illuminate\Support\Collection;
+use Mockery;
+use Symfony\Component\HttpFoundation\FileBag;
+use Symfony\Component\HttpKernel\Exception\HttpException;
+use Tests\TestCase;
+
+class PublisherControllerTest extends TestCase
+{
+    protected $publisherRepository, $publisherControllerTest, $user, $export;
+
+    public function setUp(): void
+    {
+        parent::setUp();
+        $this->publisherRepository = Mockery::mock(PublisherRepositoryInterface::class)->makePartial();
+        $this->publisherControllerTest = new PublisherController($this->publisherRepository);
+        $this->userAuthorized = factory(User::class)->make();
+        $this->userUnauthorized = new User([
+            'name' => 'Unauthorized',
+            'email' => 'Unauthorized@gmail.com',
+            'password' => '123456',
+            'address' => 'Hà Đông',
+            'phone' => '0334736187',
+            'role_id' => 10,
+            'times' => 0,
+            'status' => 0,
+        ]);
+        $this->data = [
+            'name' => 'test case 1 ne',
+            'image' => UploadedFile::fake()->image('avatar.jpg', '100', '100')->size(100),
+            'email' => 'testcase123@gmail.com',
+            'phone' => '0334736187',
+            'address' => 'HĐ',
+            'description' => '',
+        ];
+    }
+
+    public function tearDown(): void
+    {
+        Mockery::close();
+        unset($this->publisherControllerTest);
+        parent::tearDown();
+    }
+
+    public function test_index_method_authorized()
+    {
+        $this->be($this->userAuthorized);
+        $this->publisherRepository->shouldReceive('getAll');
+        $view = $this->publisherControllerTest->index();
+        $this->assertEquals('admin.publisher.index', $view->getName());
+        $this->assertArrayHasKey('publishers', $view->getData());
+    }
+
+    public function test_index_method_unauthorized()
+    {
+        $this->be($this->userUnauthorized);
+        $this->expectException(HttpException::class);
+        $this->publisherRepository->shouldReceive('getAll');
+        $view = $this->publisherControllerTest->index();
+        $this->assertEquals('admin.publisher.index', $view->getName());
+        $this->assertArrayHasKey('publishers', $view->getData());
+    }
```

Comment:
giải thích function này

Reply:
Trong hàm này
- This->be: em sẽ auth cho userUnauthorized em đã khởi tạo ở setUp
- expectException:
   - Em chưa tìm hiểu sâu về phần này, em có tìm hiểu qua là sử dụng như try catch(Exception) nhưng thay vì viêt try catch thì em viết như này
   - Em sử dụng cái này vì bên controller khi không có quyền sẽ nhảy vào abort(403) nên khi test nó sẽ ra lỗi -> em dùng cái này để bắt Exception
- shouldReceive: chỉ định hàm để mock
- Tiếp theo em gọi function của controller
- Cuối cùng là so sánh đường dẫn và data